intelhex/__init__.py

```diff
@@ -630,26 +659,48 @@ def write_hex_file(self, f, write_start_addr=True, eolstyle='native', byte_count
             minaddr = addresses[0]
             maxaddr = addresses[-1]
 
+            if extaddr_rectyp == -1:
+                # set Extended Address record type only when
+                # still not resolved - 'auto' mode and no Start Address
+                if maxaddr > 0x0FFFFF:
+                    # when Max Address larger >= 1MB
+                    extaddr_rectyp = 4
+                else:
+                    extaddr_rectyp = 2
+
             if maxaddr > 65535:
                 need_offset_record = True
             else:
                 need_offset_record = False
-            high_ofs = 0
 
+            high_ofs = 0
             cur_addr = minaddr
             cur_ix = 0
 
             while cur_addr <= maxaddr:
                 if need_offset_record:
+                    if cur_addr > 0x0FFFFF and extaddr_rectyp == 2:
```

Comment:
we should check this condition at the start of the operation. checking it here every time is waste of time.
if maxaddr > 0x0FFFFF

Reply:
OK, agree. Need to find different suitable exception type :)